tokio-net/src/tcp/stream.rs

```diff
@@ -96,23 +81,37 @@ impl TcpStream {
     ///     Ok(())
     /// }
     /// ```
-    pub fn connect(addr: &SocketAddr) -> impl Future<Output = io::Result<TcpStream>> {
-        use self::ConnectFutureState::*;
+    pub async fn connect(addr: &SocketAddr) -> io::Result<TcpStream> {
+        let sys = mio::net::TcpStream::connect(addr)?;
+        let stream = TcpStream::new(sys);
 
-        let inner = match mio::net::TcpStream::connect(addr) {
-            Ok(tcp) => Waiting(TcpStream::new(tcp)),
-            Err(e) => Error(e),
-        };
+        stream.finish_connect().await?;
+
+        Ok(stream)
+    }
+
+    async fn finish_connect(&self) -> io::Result<()> {
```

Comment:
It probably can just inlined in connect?
I don't think you'd need to use it outside of connect

Reply:
Generators are not very effective right now (size wise), so if we don't really need separate async fn, might be better to not use it